File: src/stack.rs
use std::collections::VecDeque;

#[derive(Debug)]
pub struct Node(pub f64);

#[derive(Debug)]
pub struct Stack(pub VecDeque<Node>);

impl Stack {
    pub fn new() -> Stack {
        Stack(VecDeque::new())
    }

    pub fn push(&mut self, node: Node) {
        self.push(node);
    }

    pub fn pop(&mut self) -> Option<Node> {
        self.pop()
    }

    pub fn len(&self) -> usize {
        self.len()
    }

    pub fn is_empty(&self) -> bool {
        self.is_empty()
    }
}

use std::convert::From;

impl From<f64> for Node {
    fn from(item: f64) -> Self {
        Node(item)
    }
}

File: ocaml-impl/my_own_forth/lib/stack.ml
type stack = float list

let empty () : stack = []
let is_empty (s : stack) : bool = s = []
let push (x : float) (s : stack) : stack = x :: s

let pop (s : stack) : float * stack =
  match s with
  | [] -> failwith "Stack is empty"
  | x :: xs -> x, xs
;;

let peek (s : stack) : float =
  match s with
  | [] -> failwith "Stack is empty"
  | x :: _ -> x
;;
